More AtCore Slots and Signals

sithlord48 opened the pull request:
Add all of  the test gui commands as slots within AtCore. 
Create signal for the plugins to return "PrinterStatus" data with temps.
Improve the pause and resume by tracking the heads placment when pausing and restoring it when resuming the printjob. 
setPrinterSpeed, setFlowRate, and move are now added also. 
Update the testgui to use new AtCore Features

sithlord48 committed:
+ setExtruderTemp , setBedTemp, setFanSpeed and home slots
+ AXIS enum
Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

sithlord48 committed:
Use the new slots in AtCore

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

sithlord48 committed:
PrinterStatus made into type
Plugins now Emit printerStatusChanged when durring extractTemp

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

sithlord48 committed:
Automatic style fix

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

sithlord48 committed:
Add printerStatusChanged signal to emit for temp changes

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

sithlord48 committed:
AtCore connects to plugin printerStatusChanged signal.
+ AtCore::printerStatusChanged signal to send event on to AtCore's client for logging

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

sithlord48 committed:
Connect to AtCore::printerStatusChanged and log recv of temp changes

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

lays147 commented on src/atcore.cpp:
```diff
@@ -324,3 +324,35 @@ void AtCore::detectFirmware()
+    axis.replace(QChar::fromLatin1('X'), QStringLiteral("X0"));
+    axis.replace(QChar::fromLatin1('Y'), QStringLiteral("Y0"));
+    axis.replace(QChar::fromLatin1('Z'), QStringLiteral("Z0"));
+    pushCommand(GCode::toCommand((GCode::G28), axis));
```
I don't think that you need the extra parenthesis here. only GCode::toCommand(GCode::G28, axis));

sithlord48 replied:
:heavy_check_mark:  done

lays147 commented on src/atcore.cpp:
```diff
@@ -324,3 +324,35 @@ void AtCore::detectFirmware()
+
+void AtCore::setExtruderTemp(double temp, int extruder)
+{
+    pushCommand(GCode::toCommand(GCode::M104, QString::number(extruder), QString::number(temp)));
```
Also, don't need to convert to number, since last PR from Patrick use QSring directly to form the QStringLiteral.

sithlord48 committed:
+ AtCore::pause(): to pause the print and store the location we have paused at.
+ AtCore::resume(): resume paused printjob from the location where it stoped.
+ AtCore::setAbsolutePosition(): set printer to Absolute positioning mode
+ AtCore::setRelativePosition(): set printer to Relative positioning mode

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

sithlord48 committed:
User pause and resume to pause and resume prints.
 move the head up and out of the way on pause.

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

patrickelectric commented on src/atcore.cpp:
```diff
@@ -324,3 +324,31 @@ void AtCore::detectFirmware()
+
+void AtCore::setExtruderTemp(double temp, int extruder)
+{
+    pushCommand(GCode::toCommand(GCode::M104, QString::number(extruder), QString::number(temp)));
```
temp and extruder must be uint

sithlord48 replied:
:heavy_check_mark:  done

patrickelectric commented on src/atcore.cpp:
```diff
@@ -324,3 +324,31 @@ void AtCore::detectFirmware()
+    pushCommand(GCode::toCommand(GCode::M104, QString::number(extruder), QString::number(temp)));
+}
+
+void AtCore::setBedTemp(double temp)
```
temp must be uint

sithlord48 replied:
:heavy_check_mark:  done

patrickelectric commented on src/atcore.cpp:
```diff
@@ -324,3 +324,31 @@ void AtCore::detectFirmware()
+    pushCommand(GCode::toCommand(GCode::M140, QString::number(temp)));
+}
+
+void AtCore::setFanSpeed(int speed, int fanNumber)
```
uint, uint

sithlord48 replied:
:heavy_check_mark:  done

patrickelectric commented on src/atcore.cpp:
```diff
@@ -324,3 +324,31 @@ void AtCore::detectFirmware()
+    axis.replace(QChar::fromLatin1('Y'), QStringLiteral("Y0"));
+    axis.replace(QChar::fromLatin1('Z'), QStringLiteral("Z0"));
+    pushCommand(GCode::toCommand(GCode::G28, axis));
+}
```
This function can use other function called `move(axis, pos)`, in that way we can have
a generic move function and a home function that is very simple, just calling a move

sithlord48 replied:
Home calls the home command. that way you can home xy at the same time. There are two other commands for move that are currently not implimented (yet).

patrickelectric replied:
Using a string isn't the best way to perform that, maybe:

```
printer{
    axis{
        uchar X = 1<<0;
        uchar Y = 1<<1;
        uchar Z = 1<<2;
        uchar E = 1<<3;
    }
}
```

And use the home or move function like:

```
move(printer::X, 12);
move(printer::X | printer::Z, 12);
home(printer::X | printer::Z); // Will not work with E
```

sithlord48 replied:
:heavy_check_mark:  done

patrickelectric commented on src/ifirmware.h:
```diff
@@ -3,6 +3,14 @@
 #include <QObject>
 #include <QString>
 
+struct PrinterStatus {
```
Why moving this out of IFirmware ? Global structs isn't a good idea;

sithlord48 replied:
because this structure is useful outside of iFirmware. If not as a struct perhaps we should make it a type.

patrickelectric replied:
We can expand this struct to add a better user experience with axis, status and other things:

```
printer{
    axis{
        uchar X = 1<<0;
        uchar Y = 1<<1;
        uchar Z = 1<<2;
        uchar E = 1<<3;
    }
    X{
        uint position = 0;
    }
    E{
        uint temperature = 0;   
    }
    uint timeOn;
    uint timePrinting;
    uint timeLeft
}
```

A better way to perform that is to use as a singleton.

patrickelectric commented on testclient/mainwindow.cpp:
```diff
@@ -130,6 +131,16 @@ void MainWindow::checkPushedCommands(QByteArray bmsg)
+                  .arg(QString::number(status.bedTargetTemp))
+                  .arg(QString::number(status.extruderTemp))
+                  .arg(QString::number(status.extruderTargetTemp));
+    addRLog(msg);
```
Add x,y,z position

sithlord48 replied:
x/y/z is not updated at the same time as the status and a command has to be sent to obtain the coordanates.

patrickelectric replied:
Thats true, we need to add some timers to get status of the printer in times, future PR.

patrickelectric commented on src/atcore.cpp:
```diff
@@ -150,8 +151,22 @@ QList<QSerialPortInfo> AtCore::serialPorts() const
 }
+void AtCore::setReletivePosition()
+{
+    pushCommand(QString::fromLatin1("G91"));
```
use Gcode namespace

sithlord48 replied:
:heavy_check_mark:  done

patrickelectric commented on src/atcore.cpp:
```diff
@@ -150,8 +151,22 @@ QList<QSerialPortInfo> AtCore::serialPorts() const
+
+void AtCore::setAbsolutePosition()
+{
+    pushCommand(QString::fromLatin1("G90"));
```
use Gcode namespace

sithlord48 replied:
:heavy_check_mark: Done

patrickelectric commented on testclient/mainwindow.cpp:
```diff
@@ -274,12 +274,17 @@ void MainWindow::printPBClicked()
+        core->setReletivePosition();
+        core->pushCommand(GCode::toCommand(GCode::G1, QString::fromLatin1("Z1")));
+        core->setAbsolutePosition();
+        core->pushCommand(QString::fromLatin1("G0 X0 Y195"));
```
use Gcode namespace

sithlord48 replied:
G0 is not implimented yet.

sithlord48 replied:
:heavy_check_mark: added and Done

patrickelectric commented on testclient/mainwindow.cpp:
```diff
@@ -194,37 +194,37 @@ void MainWindow::sendPBClicked()
 {
     addSLog(GCode::toString(GCode::M104));
-    core->pushCommand(GCode::toCommand(GCode::M104, ui->extTempSelCB->currentText().at(9), ui->extTempSB->cleanText()));
+    core->setExtruderTemp(ui->extTempSB->value(), QString(ui->extTempSelCB->currentText().at(9)).toInt());
```
Do not need to cast to QString

patrickelectric replied:
Ok, now I getting what are you doing.
Use QComboBox::currentIndex() and not this.

patrickelectric commented on testclient/mainwindow.cpp:
```diff
@@ -236,7 +236,7 @@ void MainWindow::mvAxisPBClicked()
 {
     addSLog(GCode::toString(GCode::M106));
-    core->pushCommand(GCode::toCommand(GCode::M106, ui->fanSpeedSelCB->currentText().at(4), ui->fanSpeedSB->cleanText()));
+    core->setFanSpeed(ui->fanSpeedSB->value(), QString(ui->fanSpeedSelCB->currentText().at(4)).toInt());
```
Do not need to cast to QString

sithlord48 replied:
We seam to need to have to use QString to convert the pulled QChar from the QComboBox to an int.

patrickelectric replied:
Use QComboBox::currentIndex()

patrickelectric commented on src/plugins/aprinterplugin.cpp:
```diff
@@ -24,16 +24,15 @@ void AprinterPlugin::extractTemp(const QString &lastMessage)
+    printerStatus.bedTargetTemp = list[3].mid(1).toFloat();
 
-    qCDebug(APRINTER_PLUGIN) << "Extruder" << PrinterStatus.extruderTemp << "/" << PrinterStatus.extruderTargetTemp;
-    qCDebug(APRINTER_PLUGIN) << "Bed" << PrinterStatus.bedTemp << "/" << PrinterStatus.bedTargetTemp;
```
Where are this debug messages ?

sithlord48 replied:
Removed since they are redundant now do we need to see 
T: TEMP
Extruder TEMP: 
Bed : Temp
Bed T/T Extruder T/T 
every second for every heat up ?

patrickelectric commented on src/gcodecommands.cpp:
```diff
@@ -91,6 +91,13 @@ QString toString(GCommands gcode)
+        } else {
+            return QStringLiteral("G0 %1").arg(value1.toUpper());
+        }
+    }
```
Split this commit in two, with one to style corrections and other to add G0.

sithlord48 replied:
:heavy_check_mark:  Done

patrickelectric commented on src/gcodecommands.cpp:
```diff
@@ -91,6 +91,13 @@ QString toString(GCommands gcode)
+        if (value1.isEmpty()) {
+            return QStringLiteral("G0");
+        } else {
+            return QStringLiteral("G0 %1").arg(value1.toUpper());
```
Check if the argument is valid, if not scream something about it in qdebug.
value must have Xnnn or Ynnn [Link](http://reprap.org/wiki/G-code#G0_.26_G1:_Move)

sithlord48 replied:
This functions is a copy of G1 . we can fix that when we fix G1.

patrickelectric replied:
Agree.

patrickelectric replied:
Added in phabricator.

patrickelectric commented on src/atcore.cpp:
```diff
@@ -324,6 +324,10 @@ void AtCore::detectFirmware()
+{
+    emit(printerStatusChanged(newStatus));
+}
 /*~~~~~Control Slots ~~~~~~~~*/
```
Space between function and block [330,331 \n].

sithlord48 replied:
:heavy_check_mark:  done

patrickelectric commented on src/atcore.cpp:
```diff
@@ -150,8 +151,22 @@ QList<QSerialPortInfo> AtCore::serialPorts() const
+        posString = message;
+        posString.resize(posString.indexOf("E"));
+        posString.replace(":", "");
+    }
```
What this is doing ? I can't get it.

sithlord48 replied:
the returned string from the printer is X:### Y:### Z:### E:#### the above is removing the E section since we do not care about the extruders position. Then we remove the remaining ":" so the args are ready to be sent later when resumed.

patrickelectric replied:
So we can run a checkMsgForStatus(message) to update the printer status inside this function, it's make more sense that let this code inside this function. 
It's better to add this last comment as a code comment to this lines.

```
the returned string from the printer is X:### Y:### Z:### E:#### the above 
is removing the E section since we do not care about the extruders position. 
Then we remove the remaining ":"
```

sithlord48 replied:
I have started to add checks within our normal checkRecvMessage functions i think that is better then having several that way at least all the logic for checking the message is in the same place.. (btw the only way to get that string is to send M114)

patrickelectric commented on src/atcore.cpp:
```diff
@@ -150,8 +151,22 @@ QList<QSerialPortInfo> AtCore::serialPorts() const
+    }
     emit(receivedMessage(lastMessage));
 }
+void AtCore::setReletivePosition()
```
Space

sithlord48 replied:
:heavy_check_mark:  done

patrickelectric commented on src/atcore.cpp:
```diff
@@ -328,6 +343,17 @@ void AtCore::statusChanged(const PrinterStatus &newStatus)
+    pushCommand(GCode::toCommand(GCode::G0, QString::fromLatin1(posString)));
+    setState(BUSY);
+}
 /*~~~~~Control Slots ~~~~~~~~*/
```
Space betwen block and function

patrickelectric commented on testclient/mainwindow.cpp:
```diff
@@ -274,12 +274,17 @@ void MainWindow::printPBClicked()
+
     default:
         qDebug() << "ERROR / STOP unhandled.";
     }
```
Commit must start with capital.

sithlord48 replied:
:heavy_check_mark:  done

sithlord48 committed:
Add G0 to GCode::toCommand

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

patrickelectric commented on src/atcore.h:
```diff
@@ -20,7 +22,12 @@ enum PrinterState {
+    Z = 1 << 2,
+    E = 1 << 3,
+};
 class AtCore : public QObject
```
Space betwen struct and class.

patrickelectric commented on testclient/mainwindow.cpp:
```diff
@@ -263,12 +274,17 @@ void MainWindow::printPBClicked()
+        core->setReletivePosition();
+        core->pushCommand(GCode::toCommand(GCode::G1, QString::fromLatin1("Z1")));
+        core->setAbsolutePosition();
+        core->pushCommand(GCode::toCommand(GCode::G0, QString::fromLatin1("X0 Y195")));
```
Why this X,Y,Z values ?

lays147 commented on src/atcore.h:
```diff
@@ -125,6 +137,50 @@ public slots:
+    void setFanSpeed(uint speed = 0, uint fanNumber = 0);
+
+    void setAbsolutePosition();
+    void setReletivePosition();
```
It's relative pos: http://reprap.org/wiki/G-code#M83:_Set_extruder_to_relative_mode

sithlord48 committed:
+ setPrinterSpeed(uint) , move(uchar,uint) and setFlowRate(uint) have been added as atcore slots

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>

sithlord48 committed:
Use the new move in atcore

Signed-off-by: Chris Rizzitello <sithlord48@gmail.com>
